Added Rendering to generate a circle

MasonScarbro opened the pull request:
This is the code I am using in order to render a circle Tied to: https://github.com/FreneticLLC/Voxalia/pull/272

MasonScarbro committed:
Added Rendering to generate a cirlce

This is the code I am using in order to render a circle

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,81 @@ public void RenderRectangle3D(Matrix4 mat)
+        bool enableDepthTest = true,
+        bool depthMask = true,
+        DepthFunction depthFunc = DepthFunction.Lequal,
+        bool enableBlending = true,
```
you don't want this blend/cull/etc. stuff in here, the caller should modify when needed. Also this should not be a `Ground` circle, just a generic circle. Engine code should be generic.

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,81 @@ public void RenderRectangle3D(Matrix4 mat)
+        Textures.White.Bind();
+
+        
+        SetColor(color == default ? new Vector4(1.0f, 0.0f, 0.0f, 1.0f) : color, view);
```
color shouldn't be here at all, but ftr `color == default` isn't right - `default` for color is `(0, 0, 0, 0)` aka standard transparent.
Also there's `Color4F` that's favored instead of `Vector4`

MasonScarbro committed:
added recommended fixes

- removed the optional params and their doc comments
Questions:
Should the the GL state alterations also be up to caller of the function to keep track of? (like should the caller need to set those out of the call and reset them after)

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,58 @@ public void RenderRectangle3D(Matrix4 mat)
+    public void RenderGroundCircle(Location center, float radius, View3D view)
+    {
+        // Save current states prior to render
+        bool cullFaceEnabled = GL.IsEnabled(EnableCap.CullFace);
```
re `Should the the GL state alterations also be up to caller of the function to keep track of? (like should the caller need to set those out of the call and reset them after)`
yes that's the caller's job to configure

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,58 @@ public void RenderRectangle3D(Matrix4 mat)
+    /// <param name="center">The Location for the circle to be rendered</param>
+    /// <param name="radius">Radius of the circle</param>
+    /// <param name="view">View to render Circle in</param>
+    public void RenderGroundCircle(Location center, float radius, View3D view)
```
this is not a `ground`circle function, this is a circle function

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,58 @@ public void RenderRectangle3D(Matrix4 mat)
+    /// <summary>
+    /// Renders a Circle based on Location (Vectors Normalized Down)
+    /// </summary>
+    /// <param name="center">The Location for the circle to be rendered</param>
```
docs should have a `.` at the end

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,58 @@ public void RenderRectangle3D(Matrix4 mat)
+
+
+    /// <summary>
+    /// Renders a Circle based on Location (Vectors Normalized Down)
```
the capitalization is weird here

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,58 @@ public void RenderRectangle3D(Matrix4 mat)
+        Textures.White.Bind();
+
+        
+       
```
clean up all this stray spacing

MasonScarbro committed:
Cleaning up + proper GL management fix

- removed all the GL state management and moved it to the called function in playerentity
- renamed function to better represent purpose
- fixed handwriting habit in code docs

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -19,6 +19,7 @@
 using FGEGraphics.GraphicsHelpers.Shaders;
 using FGEGraphics.GraphicsHelpers.Textures;
 using OpenTK;
+using OpenTK.Compute.OpenCL;
```
stray import

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -51,6 +53,9 @@ public void Init()
     /// <summary>A 3D box from (-1,-1,-1) to (1,1,1), ie size is (2,2,2) but the box is centered at (0,0,0).</summary>
     public Renderable Box;
 
+    /// <summary>A 2D circle <summary>
```
doc is wonk

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -155,6 +160,40 @@ void GenerateBoxVBO()
         Box = builder.Generate();
     }
 
+
```
missing doc

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -155,6 +160,40 @@ void GenerateBoxVBO()
+        const int segments = 32;
+        const int vertexCount = segments + 2;
+        const int indexCount = segments + 2;
+
```
stray newlines in here

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,26 @@ public void RenderRectangle3D(Matrix4 mat)
+        GL.ActiveTexture(TextureUnit.Texture0);
+        Textures.White.Bind();
+        Matrix4d mat = Matrix4d.Scale(radius, radius, 1.0)
+            * Matrix4d.CreateTranslation(center.ToOpenTK3D());
```
this shouldn't have the line cut

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,26 @@ public void RenderRectangle3D(Matrix4 mat)
     }
 
+
+
```
stray newlines here

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,26 @@ public void RenderRectangle3D(Matrix4 mat)
+    public void RenderCircle(Location center, float radius, View3D view)
+    {
+        GL.ActiveTexture(TextureUnit.Texture0);
+        Textures.White.Bind();
```
this shouldn't hardbind a white texture

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -365,6 +404,26 @@ public void RenderRectangle3D(Matrix4 mat)
+        view.SetMatrix(2, mat);
+        GL.BindVertexArray(Circle.Internal.VAO);
+        GL.DrawElements(PrimitiveType.TriangleFan, 34, DrawElementsType.UnsignedInt, IntPtr.Zero);
+    }
```
needs a `GL.BindVertexArray(0);`

MasonScarbro committed:
Fixes

- Added bind
- removed line cut
- removed stray spacing
- fixed doc
- removed stray import

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -155,6 +159,39 @@ void GenerateBoxVBO()
+            float angle = (float)(i * 2.0 * Math.PI / segments);
+            float x = (float)Math.Cos(angle);
+            float y = (float)Math.Sin(angle);
+
```
must i comment every stray newline one by one? Just clean them up

MasonScarbro replied:
I will they are hard to find I have never written code so jumbled up, so I am having trouble seeing it.

mcmonkey4eva replied:
it helps to look at the PR change view of your own work, so you're only looking at the changed lines

mcmonkey4eva commented on FGEGraphics/GraphicsHelpers/Renderer.cs:
```diff
@@ -155,6 +159,39 @@ void GenerateBoxVBO()
+        builder.Tangents[0] = new Vector3(1, 0, 0);
+        builder.Indices[0] = 0;
+        // Generate vertices around circle
+        for (int i = 0; i <= segments; i++)
```
I believe only the indices should have 1 extra, not the vertices -- there's a center, and 32 vertices, only the index needs to repeat `1` at the end to close the loop

MasonScarbro committed:
fixed segment thing I think?

MasonScarbro committed:
reversed circle vao

- This should fix it but should there be a way to turn it back without cullfacing it? Overloaded function or optional param?
